network/p2p/p2p.go

```diff
@@ -475,3 +825,67 @@ func (n *p2pNetwork) getMaxPeers(topic string) int {
 	}
 	return n.cfg.TopicMaxPeers
 }
+
+// peerScore calculates a score for peerID based on how valuable this peer's contribution
+// to us assessing each subnet-contribution he makes (as estimated by score func).
+func (n *p2pNetwork) peerScore(peerID peer.ID) float64 {
+	result := 0.0
+
+	peerSubnets := n.idx.GetPeerSubnets(peerID)
+	sharedSubnets := commons.SharedSubnets(n.activeSubnets, peerSubnets, 0)
+	for _, subnet := range sharedSubnets {
+		result += n.score(peerID, subnet)
+	}
+
+	return result
+}
+
+// score assesses how valuable the contribution of peerID to specified subnet is by calculating
+// how valuable this peer would have been if we didn't have him, but then connected with.
+func (n *p2pNetwork) score(peerID peer.ID, subnet int) float64 {
+	filterOutPeer := func(peerID peer.ID, peerIDs []peer.ID) []peer.ID {
+		if len(peerIDs) == 0 {
+			return nil
+		}
+		result := make([]peer.ID, 0, len(peerIDs))
+		for _, elem := range peerIDs {
+			if elem == peerID {
+				continue
+			}
+			result = append(result, elem)
+		}
+		return result
+	}
+
+	topic := strconv.Itoa(subnet)
+	subnetPeers, err := n.topicsCtrl.Peers(topic)
+	if err != nil {
+		n.interfaceLogger.Debug(
+			"cannot score peer with respect to this subnet, assuming zero contribution",
+			zap.String("topic", topic),
+			zap.Error(fmt.Errorf("could not get topic peers: %w", err)),
+		)
+		return 0.0
+	}
+	subnetPeersExcluding := len(filterOutPeer(peerID, subnetPeers))
+
+	const targetPeersPerSubnet = 3
+	return score(targetPeersPerSubnet, subnetPeersExcluding)
+}
```

Comment:
is this not equivalent?
```suggestion
// score assesses how valuable the contribution of peerID to specified subnet is by calculating
// how valuable this peer would have been if we didn't have him, but then connected with.
func (n *p2pNetwork) score(peerID peer.ID, subnet int) float64 {
	topic := strconv.Itoa(subnet)
	subnetPeers, err := n.topicsCtrl.Peers(topic)
	if err != nil {
		n.interfaceLogger.Debug(
			"cannot score peer with respect to this subnet, assuming zero contribution",
			zap.String("topic", topic),
			zap.Error(fmt.Errorf("could not get topic peers: %w", err)),
		)
		return 0.0
	}
	subnetPeersExcluding := 0
    for _, p := range subnetPeers {
        if p != peerID {
            subnetPeersExcluding++
        }
    }

	const targetPeersPerSubnet = 3
	return score(targetPeersPerSubnet, subnetPeersExcluding)
}
```

Reply:
I think it is equivalent, but it's less readable IMO because `filterOutPeer` hides away this chunk of code that you'd need to read trough to figure out what's going on (maybe it's a matter of preference, but doesn't seem all that important either way):
```
subnetPeersExcluding := 0
    for _, p := range subnetPeers {
        if p != peerID {
            subnetPeersExcluding++
        }
    }
```